### PR TITLE
[sonic_installer] Add swap setup support

### PR DESCRIPTION
**Pull request by lolyu**
#### What I did
* Let's add swap memory setup support for `sonic_installer` command so it could run on devices with limited memory resources.

#### How I did it
* Add the following new options to `sonic_installer`:
  1. `--skip-setup-swap`: if present, will skip setup swap memory.
  2.  `--swap-mem-size`: this will change the swap memory size(the default swap size is 1024 MiB)
  3. `--total-mem-threshold`: if the system total memory is less than the value passed to `--total-mem-threshold`(default 2048 MiB), `sonic_installer` will setup swap memory.
  4.  `--available-mem-threshold`: if the system available memory is less than the value passed to `--available-mem-threshold`(default 1200 MiB), `sonic_installer` will setup swap memory.
* Add class `MutuallyExclusiveOption` to check the mutually-exclusive relationship between options.
* Add class `SWAPAllocator` to support swap memory setup/remove functionalities.
* Also, when `sonic_installer` tries to setup swap, if the system disk free space is less than 4096 MiB, `sonic_installer` will not setup swap memory. 

#### How to verify it
Run `sonic_installer` over devices with limited memory
* `sonic_installer install --help` output:
```
# sonic_installer install --help
Warning: 'sonic_installer' command is deprecated and will be removed in the future
Please use 'sonic-installer' instead
Usage: sonic_installer install [OPTIONS] URL

  Install image from local binary or URL

Options:
  -y, --yes
  -f, --force                     Force installation of an image of a type
                                  which differs from that of the current
                                  running image
  --skip_migration                Do not migrate current configuration to the
                                  newly installed image
  --skip-package-migration        Do not migrate current packages to the newly
                                  installed image
  --skip-setup-swap               Skip setup temporary SWAP memory used for
                                  installation
  --swap-mem-size INTEGER         SWAP memory space size  [default: (1024
                                  MiB)] NOTE: this argument is mutually
                                  exclusive with arguments: skip_setup_swap
  --total-mem-threshold INTEGER   If system total memory is lower than
                                  threshold, setup SWAP memory  [default:
                                  (2048 MiB)] NOTE: this argument is mutually
                                  exclusive with arguments: skip_setup_swap
  --available-mem-threshold INTEGER
                                  If system available memory is lower than
                                  threhold, setup SWAP memory  [default: (1200
                                  MiB)] NOTE: this argument is mutually
                                  exclusive with arguments: skip_setup_swap
  --help                          Show this message and exit.
```

### CODE REVIEW

File: sonic_installer/main.py [hunk 1]
```diff
@@ -367,6 +368,101 @@ def migrate_sonic_packages(bootloader, binary_image_version):
...
+
+    @staticmethod
+    def read_from_meminfo():
+        """Read information from /proc/meminfo."""
+        meminfo = {}
+        with open("/proc/meminfo") as fd:
+            for line in fd.readlines():
+                if line:
+                    fields = line.split()
+                    meminfo[fields[0].rstrip(":")] = int(fields[1])
```
Comment (yxieca) [inline sonic_installer/main.py:415]: Protect against less than 2 fields cases and integer conversion failures?

File: sonic_installer/main.py [hunk 2]
```diff
@@ -367,6 +368,101 @@ def migrate_sonic_packages(bootloader, binary_image_version):
...
+        return meminfo
+
+    def setup_swapmem(self):
+        swapfile = SWAPAllocator.SWAP_FILE_PATH
+        with open(swapfile, 'wb') as fd:
+            os.posix_fallocate(fd.fileno(), 0, self.swap_mem_size * SWAPAllocator.MiB_TO_BYTES_FACTOR)
```
Comment (yxieca) [inline sonic_installer/main.py:421]: Would this fail if the file exists?

File: sonic_installer/main.py [hunk 3]
```diff
@@ -367,6 +368,101 @@ def migrate_sonic_packages(bootloader, binary_image_version):
...
+        try:
+            os.unlink(swapfile)
+        finally:
+            pass
+
+    def __enter__(self):
+        if self.allocate:
+            if self.get_disk_freespace('/host') < max(SWAPAllocator.DISK_FREESPACE_THRESHOLD, self.swap_mem_size) * SWAPAllocator.MiB_TO_BYTES_FACTOR:
+                echo_and_log("Failed to setup SWAP memory due to insufficient disk free space.\nAborting...", LOG_ERR)
+                raise click.Abort()
```
Comment (yxieca) [inline sonic_installer/main.py:437]: This shouldn't be a failure condition. There is not enough hd to setup swap, we should proceed with installing.

Comment (yxieca) [reply sonic_installer/main.py:437]: You could add disk free space in the error log to help debug.

File: sonic_installer/main.py [hunk 4]
```diff
@@ -389,8 +485,22 @@ def sonic_installer():
...
+              cls=clicommon.SubOrdinateOption, depends_on=['setup_swap_if_necessary'],
+              callback=validate_positive_int)
+@click.option('--available-mem-threshold', default=1200, type=int, show_default='1200 MiB',
+              help='If system available memory is lower than threhold, setup SWAP memory',
+              cls=clicommon.SubOrdinateOption, depends_on=['setup_swap_if_necessary'],
+              callback=validate_positive_int)
 @click.argument('url')
-def install(url, force, skip_migration=False, skip_package_migration=False):
+def install(url, force, skip_migration=False, skip_package_migration=False,
+            setup_swap_if_necessary=False, swap_mem_size=None, total_mem_threshold=None, available_mem_threshold=None):
```
Comment (yxieca) [inline sonic_installer/main.py:503]: Our default behavior should be trying to setup this swap mem when needed. You can add an option to disable the behavior.

Comment (lolyu) [reply sonic_installer/main.py:503]: Sure, will change.

File: sonic_installer/main.py (hunk 2 above)
Comment (lolyu) [reply sonic_installer/main.py:421]: Line 445~446 will ensure the swapfile is removed before calling `setup_swapmem`.

File: sonic_installer/main.py (hunk 1 above)
Comment (lolyu) [reply sonic_installer/main.py:415]: Sure, this could be safe-guarded.

**Comment by lolyu** [comment]
```
tests/swap_allocator_test.py::TestSWAPAllocator::test_read_from_meminfo PASSED                                                                                                                 [  7%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_setup_swapmem PASSED                                                                                                                     [ 15%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_remove_swapmem PASSED                                                                                                                    [ 23%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_initialization_default_args PASSED                                                                                        [ 30%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_initialization_custom_args PASSED                                                                                         [ 38%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_allocate_true_insufficient_total_memory PASSED                                                              [ 46%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_allocate_true_insufficient_available_memory PASSED                                                          [ 53%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_allocate_true_insufficient_disk_space PASSED                                                                [ 61%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_allocate_true_swapfile_present PASSED                                                                       [ 69%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_setup_error PASSED                                                                                          [ 76%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_enter_allocate_false PASSED                                                                                       [ 84%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_exit_is_allocated_true PASSED                                                                                     [ 92%]
tests/swap_allocator_test.py::TestSWAPAllocator::test_swap_allocator_context_exit_is_allocated_false PASSED                                                                                    [100%]
```

**Comment by lolyu** [comment]
@yxieca, could you please take a review for my updates? Thanks!

**Review by yxieca — APPROVED** [review]

**Comment by lolyu** [comment]
Looks like this PR could not be cherry-picked to 202012, will create a separate PR for 202012 branch.

**Comment by qiluo-msft** [comment]
This commit could not be cleanly cherry-pick to 202012. Please submit another PR.

**Comment by judyjoseph** [comment]
This commit could not be cleanly cherry-pick to 202106. Please submit another PR.

**Pull request merged**